titus-server-master/src/main/java/com/netflix/titus/master/scheduler/TaskPlacementRecorder.java

```diff
@@ -207,6 +220,23 @@ private void removeTask(TaskAssignmentResult assignmentResult, TitusQueuableTask
         }
     }
 
+    private Map<String, String> buildOpportunisticResourcesContext(VirtualMachineLease lease, TitusQueuableTask fenzoTask) {
+        int count = fenzoTask.getOpportunisticCpus();
+        if (count < 1) {
+            return Collections.emptyMap();
+        }
+        Optional<OpportunisticCpuAvailability> availability = opportunisticCpuCache.findAvailableOpportunisticCpus(lease.getVMID());
+        if (!availability.isPresent()) {
+            titusRuntime.getCodeInvariants().inconsistent("Task %s was scheduled on opportunistic CPUs that are not available",
```

Comment:
I don't understand this message.  What does "unavailable" mean?  They exist but are from the wrong time period?

Reply:
This should never happen. If it happens, likely I made a mistake. Code invariants are markers we use to point to inconsistencies that need to be fixed, and the code is likely wrong somewhere.

"Not available" here means the opportunistic resource was available right before during the scheduling loop, and we can't find it here when recording the placement decision anymore. It should never happen since we freeze the state of the world (resources and machines available) during the scheduling iteration.

Reply:
ok, thanks.  makes sense